lib/create-contentful-api.ts

```diff
@@ -43,18 +45,22 @@ import resolveCircular from './utils/resolve-circular'
 import validateTimestamp from './utils/validate-timestamp'
 import {
   ChainOptions,
-  isClientWithAllLocalesAndWithLinkResolutionAndWithoutUnresolvableLinks,
-  isClientWithAllLocalesAndWithLinkResolutionAndWithUnresolvableLinks,
-  isClientWithAllLocalesAndWithoutLinkResolution,
-  isClientWithLinkResolutionAndWithoutUnresolvableLinks,
-  isClientWithLinkResolutionAndWithUnresolvableLinks,
-  isClientWithoutLinkResolution,
+  ChainOptionWithAllLocalesAndWithLinkResolutionAndWithoutUnresolvableLinks,
+  ChainOptionWithAllLocalesAndWithLinkResolutionAndWithUnresolvableLinks,
+  ChainOptionWithAllLocalesAndWithoutLinkResolution,
+  ChainOptionWithLinkResolutionAndWithoutUnresolvableLinks,
+  ChainOptionWithLinkResolutionAndWithUnresolvableLinks,
+  ChainOptionWithoutLinkResolution,
 } from './utils/client-helpers'
 import { validateLocaleParam, validateResolveLinksParam } from './utils/validate-params'
 
 const ASSET_KEY_MAX_LIFETIME = 48 * 60 * 60
 
 export interface ClientWithLinkResolutionAndWithUnresolvableLinks extends BaseClient {
+  withAllLocales: ClientWithAllLocalesAndWithLinkResolutionAndWithUnresolvableLinks
```

Comment:
Unresolvable links don’t exist when you don’t have link resolution so it doesn’t actually matter. The boolean internally has a value, of course, but it’s not read in this case.
As for the locales I’d try to avoid microoptimizations for now and rather spend the time trying to find a general approach to shorten the types like you mentioned as well.